network/stream.go

```diff
@@ -5,82 +5,89 @@ import (
 	"encoding/binary"
 	"io"
 	"net"
-	"time"
+	"sync"
 
 	"github.com/gogo/protobuf/proto"
+	"github.com/golang/glog"
 	"github.com/perlin-network/noise/crypto"
 	"github.com/perlin-network/noise/protobuf"
 	"github.com/pkg/errors"
 )
 
 // sendMessage marshals, signs and sends a message over a stream.
-func (n *Network) sendMessage(stream net.Conn, message *protobuf.Message) error {
+func (n *Network) sendMessage(w io.Writer, message *protobuf.Message, writerMutex *sync.Mutex) error {
 	bytes, err := proto.Marshal(message)
 	if err != nil {
 		return errors.Wrap(err, "failed to marshal message")
 	}
 
 	// Serialize size.
-	buffer := make([]byte, binary.MaxVarintLen64)
-	binary.PutUvarint(buffer, uint64(len(bytes)))
+	buffer := make([]byte, 4)
+	binary.BigEndian.PutUint32(buffer, uint32(len(bytes)))
 
-	// Prefix message with its size.
-	bytes = append(buffer, bytes...)
+	buffer = append(buffer, bytes...)
+	totalSize := len(buffer)
 
-	stream.SetDeadline(time.Now().Add(3 * time.Second))
+	// Write until all bytes have been written.
+	bytesWritten, totalBytesWritten := 0, 0
 
-	writer := bufio.NewWriter(stream)
+	writerMutex.Lock()
 
-	// Send request bytes.
-	written, err := writer.Write(bytes)
-	if err != nil {
-		return errors.Wrap(err, "failed to send request bytes")
+	bw, isBuffered := w.(*bufio.Writer)
+	if isBuffered && (bw.Buffered() > 0) && (bw.Available() < totalSize) {
+		if err := bw.Flush(); err != nil {
+			return err
+		}
 	}
 
-	// Flush writer.
-	err = writer.Flush()
-	if err != nil {
-		return errors.Wrap(err, "failed to flush writer")
+	for totalBytesWritten < len(buffer) && err == nil {
+		bytesWritten, err = w.Write(buffer[totalBytesWritten:])
+		if err != nil {
+			glog.Errorf("stream: failed to write entire buffer, err: %+v\n", err)
```

Comment:
Is it okay to keep writing or should we break out of the loop on error and retry from the start

Reply:
well it will break out due to the err == nil in the if check